packages/addon/lua/gelly/mods/blood-mod/init.lua

```diff
@@ -31,51 +51,93 @@ local BLOOD_CONFIGS = {
 	},
 
 	-- shotgun
-	[536870914] = {
+	weapon_shotgun = {
 		MinDensity = 100,
 		MaxDensity = 200,
-		VelocityPower = 14,
-		Randomness = 0.1,
-		CubeSize = 5,
+		VelocityPower = 5,
+		Randomness = 0.5,
+		CubeSize = 15,
+		DamageMultiplier = 35,
 	},
 
-	[DMG_SLASH] = {
+	weapon_357 = {
 		MinDensity = 100,
 		MaxDensity = 200,
-		VelocityPower = 4,
+		VelocityPower = 5,
 		Randomness = 0.1,
-		CubeSize = 3,
+		CubeSize = 5,
+		DamageMultiplier = 30,
 	},
+}
 
-	[DMG_CLUB] = {
-		MinDensity = 100,
-		MaxDensity = 200,
-		VelocityPower = 4,
-		Randomness = 0.1,
-		CubeSize = 3,
+local BLOOD_COLOR_MATERIALS = {
+	[BLOOD_COLOR_RED] = {
+		Roughness = 0,
+		IsSpecularTransmission = true,
+		RefractiveIndex = 1.373,
+		Absorption = Vector(0.05, 0.5, 0.5),
+		DiffuseColor = Vector(0, 0, 0),
 	},
-
-	[DMG_BLAST] = {
-		MinDensity = 3200,
-		MaxDensity = 5200,
-		VelocityPower = 40,
-		Randomness = 1,
-		CubeSize = 10,
-		DamageMultiplier = 90,
+	[BLOOD_COLOR_YELLOW] = {
+		Roughness = 0,
+		IsSpecularTransmission = true,
+		RefractiveIndex = 1.373,
+		Absorption = Vector(0.25, 0.3, 0.5),
+		DiffuseColor = Vector(0, 0, 0),
+	},
+	[BLOOD_COLOR_GREEN] = {
+		Roughness = 0,
+		IsSpecularTransmission = true,
+		RefractiveIndex = 1.373,
+		Absorption = Vector(0.3, 0.25, 0.5),
+		DiffuseColor = Vector(0, 0, 0),
+	},
+	[BLOOD_COLOR_ANTLION] = {
+		Roughness = 0,
+		IsSpecularTransmission = true,
+		RefractiveIndex = 1.373,
+		Absorption = Vector(0.45, 0.1, 0.5),
+		DiffuseColor = Vector(0, 0, 0),
+	},
+	[BLOOD_COLOR_ZOMBIE] = {
+		Roughness = 0,
+		IsSpecularTransmission = true,
+		RefractiveIndex = 1.373,
+		Absorption = Vector(0.2, 0.35, 0.5),
+		DiffuseColor = Vector(0, 0, 0),
+	},
+	[BLOOD_COLOR_ANTLION_WORKER] = {
+		Roughness = 0,
+		IsSpecularTransmission = true,
+		RefractiveIndex = 1.373,
+		Absorption = Vector(0.5, 0.05, 0.5),
+		DiffuseColor = Vector(0, 0, 0),
 	},
 }
 
-local function sprayBlood(damageType, attacker, position, force, damage)
-	local normal = force:GetNormalized()
-	if attacker == LocalPlayer() then
-		normal = LocalPlayer():GetAimVector()
+local function getConfig(damageType)
+	for _, config in ipairs(DAMAGE_TYPE_BLOOD_CONFIGS) do
+        if bit.band(damageType, config.DamageFlags) ~= 0 then
+            return config
+        end
 	end
+	return nil
+end
 
-	local config = BLOOD_CONFIGS[damageType]
-
-	if not config then
-		return
+local function getWeaponConfig(attacker)
+	if attacker:IsNPC() or attacker:IsPlayer() then
+		return WEAPON_BLOOD_CONFIGS[attacker:GetActiveWeapon():GetClass()]
 	end
+	return nil
+end
+
+local function sprayBlood(damageType, attacker, position, force, damage)
+	local normal = attacker == LocalPlayer() and 
+					attacker:GetAimVector() or 
+					force:GetNormalized()
+
+	local config = getWeaponConfig(attacker) or getConfig(damageType)
+	if not config then return end
```

Comment:
Couple of comments on this:
- nit from the last review: We still have the business logic of which configs take priority in the sprayBlood function. Would be cleaner to move that into its own function
- By the time we return here, we’ve already looked up and set the material in the parent’s scope
  - To resolve this, you could extract the material selection logic into a separate function, and then set and unset the material here instead without compromising readability